### ISSUE TITLE
[Issue]: [gfx1201][RDNA4][CK] FlashAttention backward remains 3–7x slower than PyTorch SDPA; profiling points to heavy spills and dQ K-chunk scaling

### ISSUE THREAD

**Issue by qweqweewqe7-create** [ROCm/flash-attention#188]
### Problem Description

<html>
<body>
<h1 data-pm-slice="1 1 []"><span>[Issue]: [gfx1201][RDNA4][CK] FlashAttention backward remains 3–7x slower than PyTorch SDPA; profiling points to heavy spills and dQ K-chunk scaling</span></h1><h2><span>Problem Description</span></h2><p><span>I am seeing severe backward-performance issues with the ROCm/Composable Kernel FlashAttention path on an AMD Radeon RX 9070 XT (</span><code dir="ltr"><span>gfx1201</span></code><span>, RDNA4) on Windows.</span></p><p><span>I first reported the end-to-end behavior in Dao-AILab/flash-attention#2800:</span><br><a href="https://github.com/Dao-AILab/flash-attention/issues/2800" data-rich-text-autolink=""><span>https://github.com/Dao-AILab/flash-attention/issues/2800</span></a></p><p><span>Since then I have gone considerably deeper: I captured RGP traces, extracted the generated gfx1201 code object, inspected resource metadata/disassembly, changed scheduling/liveness in the CK backward pipeline, and built several correctness-checked structural variants that isolate </span><code dir="ltr"><span>dQ</span></code><span>, </span><code dir="ltr"><span>dK</span></code><span>, and </span><code dir="ltr"><span>dV</span></code><span> costs.</span></p><p><span>The current evidence suggests </span><strong><span>two distinct gfx1201 backward bottlenecks</span></strong><span>:</span></p><ol data-spread="false" start="1"><li><p><strong><span>Very high VGPR pressure/spilling in the fused backward kernel.</span></strong></p></li><li><p><strong><span>The non-QrQtrDor K-major dQ organization appears to scale almost directly with the number of K chunks</span></strong><span>, consistent with repeated partial-dQ accumulation / </span><code dir="ltr"><span>dq_acc</span></code><span> workspace traffic becoming a major long-sequence cost.</span></p></li></ol><p><span>This is not a correctness failure. The normal CK path and the gradient(s) retained by the diagnostic variants remain numerically close to PyTorch SDPA.</span></p><p><span>I am filing this on the ROCm side because the issue appears to be in the AMD CK/gfx12 backend rather than the Python API layer. AMD's Radeon documentation advertises FlashAttention-2 backward for training, and the upstream FlashAttention README lists RDNA4 backward support for the CK backend with </span><code dir="ltr"><span>deterministic=False</span></code><span>, which is the mode tested here:</span></p><ul data-spread="false"><li><p><a href="https://rocm.docs.amd.com/projects/radeon/en/latest/docs/compatibility.html" data-rich-text-autolink=""><span>https://rocm.docs.amd.com/projects/radeon/en/latest/docs/compatibility.html</span></a></p></li><li><p><a href="https://github.com/Dao-AILab/flash-attention#amd-rocm-support" data-rich-text-autolink=""><span>https://github.com/Dao-AILab/flash-attention#amd-rocm-support</span></a></p></li></ul><h2><span>Environment</span></h2><ul data-spread="false"><li><p><span>OS: Windows 11, build </span><code dir="ltr"><span>26100.1</span></code></p></li><li><p><span>CPU: AMD Ryzen 7 9800X3D</span></p></li><li><p><span>GPU: AMD Radeon RX 9070 XT</span></p></li><li><p><span>Architecture: </span><code dir="ltr"><span>gfx1201</span></code><span> (RDNA4)</span></p></li><li><p><span>Driver: Adrenalin </span><code dir="ltr"><span>26.8.1</span></code><span> (</span><code dir="ltr"><span>26.10.41.01-260811a-203304C</span></code><span>)</span></p></li><li><p><span>PyTorch: </span><code dir="ltr"><span>2.15.0a0+rocm10.1.0a20260822</span></code></p></li><li><p><span>HIP: </span><code dir="ltr"><span>7.16.26332</span></code></p></li><li><p><span>FlashAttention Python package: </span><code dir="ltr"><span>2.8.4</span></code></p></li><li><p><span>FlashAttention source tree used for the investigation: </span><code dir="ltr"><span>fa4-v4.0.0.beta26</span></code></p></li><li><p><span>CK submodule pin: </span><code dir="ltr"><span>c56c6750d0fc54ed771d532cc92c316423449614</span></code></p></li><li><p><span>AMD clang: </span><code dir="ltr"><span>23.0.0git</span></code></p></li><li><p><span>MSVC toolset used for successful local builds: </span><code dir="ltr"><span>14.44.35207</span></code></p></li><li><p><span>CK backend enabled</span></p></li><li><p><span>Triton AMD backend disabled</span></p></li><li><p><code dir="ltr"><span>deterministic=False</span></code></p></li></ul><p><span>The machine also contains a second Radeon GPU, but I expose only the RX 9070 XT to the benchmark, so it is logical </span><code dir="ltr"><span>cuda:0</span></code><span>.</span></p><h2><span>Test configuration</span></h2><p><span>Unless otherwise noted:</span></p><pre dir="ltr"><code dir="ltr"><span>dtype       = bfloat16
batch       = 1
heads       = 16
head_dim    = 128
seqlen_q    = seqlen_k = 1024 / 2048 / 4096
causal      = false
dropout     = 0
bias        = none
deterministic = false</span></code></pre><h2><span>Original end-to-end behavior</span></h2><p><span>The original report used the 2026-08-14 ROCm/PyTorch nightly and an essentially stock CK path.</span></p><h3><span>Forward</span></h3>
Sequence | PyTorch SDPA | Flash CK | CK / SDPA
-- | -- | -- | --
1024 | 0.267 ms | 0.134 ms | 0.503x
2048 | 0.610 ms | 0.439 ms | 0.719x
4096 | 2.011 ms | 1.450 ms | 0.721x

<p><span>Numerically, dQ remains correct in both cases.</span></p><p><span>Halving N doubles the number of K chunks. The near-2x timing change therefore strongly suggests that the cost of the current dQ organization is tied to the number of K chunks rather than being primarily a register-footprint effect.</span></p><h3><span>3. Why I suspect partial-dQ accumulation/workspace traffic</span></h3><p><span>The non-QrQtrDor backward path is K-major: a workgroup owns a K tile/chunk and traverses Q. Each K chunk therefore produces only a partial contribution to dQ.</span></p><p><span>In the non-deterministic path, those partial dQ results are accumulated into the FP32 </span><code dir="ltr"><span>dq_acc</span></code><span> workspace with atomic-add semantics and converted later.</span></p><p><span>Conceptually:</span></p><pre dir="ltr"><code dir="ltr"><span>K chunk 0 CTA -&gt; partial dQ over Q rows --\
K chunk 1 CTA -&gt; partial dQ over Q rows ---+--&gt; dq_acc workspace / accumulation
K chunk 2 CTA -&gt; partial dQ over Q rows --/
...</span></code></pre><p><span>The N128 -&gt; N64 result is consistent with this: twice as many K chunks gives almost exactly twice the isolated dQ time at S=4096.</span></p><h3><span>4. Simply increasing N is blocked by LDS in the current layout</span></h3><p><span>I also tried N256 as a diagnostic, hoping to reduce the number of K chunks. It failed at link time because the current layout requires too much LDS:</span></p><pre dir="ltr"><code dir="ltr"><span>ld.lld: error: local memory (131072) exceeds limit (65536)</span></code></pre><p><span>So the obvious workaround of making the current K tile much larger is not available without changing the LDS organization.</span></p><h2><span>Working hypothesis / possible direction</span></h2><p><span>My current hypothesis is that gfx1201 backward has </span><strong><span>two interacting problems</span></strong><span>:</span></p><ul data-spread="false"><li><p><span>the fused dK/dV/dQ live set creates severe VGPR spilling, and</span></p></li><li><p><span>the K-major dQ scheme incurs expensive repeated partial-dQ accumulation as sequence length / K-chunk count grows.</span></p></li></ul><p><span>A potentially better gfx1201 organization may therefore be:</span></p><pre dir="ltr"><code dir="ltr"><span>Pass A: dK + dV, with a smaller-N tile / lower live-set pressure
Pass B: Q-owned dQ</span></code></pre><p><span>For the dQ pass, instead of one CTA owning a K chunk and producing partial dQ, one CTA would own a Q tile, iterate over all K chunks internally, keep its dQ accumulator local across that loop, and write dQ once at the end:</span></p><pre dir="ltr"><code dir="ltr"><span>Q tile CTA:
    dq_acc = 0
    for each K tile:
        recompute/use P and dS
        dq_acc += dS @ K^T
    write dQ once</span></code></pre><p><span>CK already has a QrQtrDor-style organization where dQ can be written directly without allocating the usual </span><code dir="ltr"><span>dq_acc</span></code><span> workspace. I have not found a way for the gfx12 path used here to select/adapt that organization for this long-Sq, noncausal training case.</span></p><p><span>I am </span><strong><span>not</span></strong><span> claiming that this split is necessarily the final optimal implementation. The diagnostic K-major </span><code dir="ltr"><span>DQ-only</span></code><span> kernel is actually too slow to use as-is: </span><code dir="ltr"><span>DKDV N64 + DQ N128</span></code><span> would be worse than the fused kernel. The important observation is the scaling behavior, which suggests that changing dQ ownership/traversal may matter more than further tuning the current K-major DQ tile.</span></p><h2><span>RGP / low-level artifacts available</span></h2><p><span>I have a successful Radeon Developer Panel instruction-tracing capture of the target backward workload on the RX 9070 XT. One representative RGP trace is about 64 MB (</span><code dir="ltr"><span>64,231,832</span></code><span> bytes).</span></p><p><span>I can provide, if useful:</span></p><ul data-spread="false"><li><p><span>the RGP instruction trace</span></p></li><li><p><span>extracted gfx1201 </span><code dir="ltr"><span>.co</span></code><span> code object</span></p></li><li><p><code dir="ltr"><span>llvm-readelf --notes</span></code><span> resource metadata</span></p></li><li><p><span>full gfx1201 disassembly</span></p></li><li><p><span>the scheduling/liveness experimental headers</span></p></li><li><p><span>DKDV-only / DK-only / DV-only / DQ-only diagnostic variants</span></p></li><li><p><span>benchmark scripts and wheel builds</span></p></li></ul><h2><span>Why I do not think this is Python overhead</span></h2><p><span>A few observations point specifically at the generated GPU backward path:</span></p><ul data-spread="false"><li><p><span>CK forward is faster than SDPA in the same Python/process/environment.</span></p></li><li><p><span>Small changes to the generated gfx1201 kernel change backward time by tens of milliseconds at S=4096.</span></p></li><li><p><span>The gradient-isolation variants show highly structured scaling rather than a fixed launch/API overhead.</span></p></li><li><p><span>DQ N64 vs N128 tracks K-chunk count almost 1:1 at long sequence length.</span></p></li><li><p><span>The generated code object explicitly reports extreme VGPR spilling/private-segment use.</span></p></li></ul><h2><span>Questions for ROCm / CK maintainers</span></h2><ol data-spread="false" start="1"><li><p><span>Is this level of spilling and 3–7x backward slowdown expected for the current gfx1201/RDNA4 CK FlashAttention path?</span></p></li><li><p><span>Is BF16, D128, non-deterministic backward on gfx1201 currently considered performance-tuned, or is it still using a relatively generic/fallback organization?</span></p></li><li><p><span>Does the near-2x dQ slowdown when </span><code dir="ltr"><span>kN0</span></code><span> is halved match an expected cost in the partial-dQ workspace/atomic path?</span></p></li><li><p><span>Is there an intended gfx12 path for a Q-owned / direct-write dQ traversal for long-Sq training, similar in spirit to QrQtrDor?</span></p></li><li><p><span>Would it make sense to specialize gfx1201 backward as separate </span><code dir="ltr"><span>DKDV + DQ</span></code><span> passes, or is there another CK design that avoids the fused live set and dQ partial accumulation without duplicating too much QK/softmax work?</span></p></li><li><p><span>If the kernel-side part of this should instead be tracked in </span><code dir="ltr"><span>ROCm/rocm-libraries</span></code><span> / Composable Kernel, please let me know and I can cross-reference or move the detailed analysis there.</span></p></li><li><p><span>Which profiling artifact would be most useful first: RGP, code object + ISA, or the diagnostic source patches?</span></p></li></ol><h2><span>Related issues / references</span></h2><ul data-spread="false"><li><p><span>Original report: Dao-AILab/flash-attention#2800</span><br><a href="https://github.com/Dao-AILab/flash-attention/issues/2800" data-rich-text-autolink=""><span>https://github.com/Dao-AILab/flash-attention/issues/2800</span></a></p></li><li><p><span>Older ROCm FlashAttention backward-performance issue on MI250X: ROCm/flash-attention#44</span><br><a href="https://github.com/ROCm/flash-attention/issues/44" data-rich-text-autolink=""><span>https://github.com/ROCm/flash-attention/issues/44</span></a></p></li><li><p><span>Earlier RDNA3+ backward discussion: ROCm/composable_kernel#1434</span><br><a href="https://github.com/ROCm/composable_kernel/issues/1434" data-rich-text-autolink=""><span>https://github.com/ROCm/composable_kernel/issues/1434</span></a></p></li><li><p><span>ROCm Radeon compatibility documentation (FlashAttention-2 backward enabled):</span><br><a href="https://rocm.docs.amd.com/projects/radeon/en/latest/docs/compatibility.html" data-rich-text-autolink=""><span>https://rocm.docs.amd.com/projects/radeon/en/latest/docs/compatibility.html</span></a></p></li><li><p><span>Upstream FlashAttention AMD ROCm support section:</span><br><a href="https://github.com/Dao-AILab/flash-attention#amd-rocm-support" data-rich-text-autolink=""><span>https://github.com/Dao-AILab/flash-attention#amd-rocm-support</span></a></p></li></ul><p><span>I am happy to attach the RGP trace and code object immediately, or to clean up one of the diagnostic variants into a smaller patch/reproducer if that is more useful for triage.</span></p>
</body>
</html>

### Operating System

Windows 11, build 26100.1

### CPU

AMD Ryzen 7 9800X3D

### GPU

AMD Radeon RX 9070 XT

### ROCm Version

rocm10.1.0a20260822

### Installation Method

Python wheels

### Installed ROCm Packages / Versions

<details>
<summary>Installed ROCm packages / versions</summary>

```
PyTorch: 2.15.0a0+rocm10.1.0a20260822

HIP: 7.16.26332
```

</details>
